File: HisabCore/Sources/HisabCore/GenericBank/FormatSpec.swift
import Foundation

/// A bank-statement layout, described declaratively. Specs are trusted for
/// format *selection* only — their output must still close the balance chain
/// on the user's own file. Hard rule: this format never grows an expression
/// language or any eval-like field.
public struct FormatSpec: Codable, Equatable, Sendable {
    public var id: String                       // "sbi-table"
    public var sourceID: String                 // "bank:sbi" (or a frozen legacy id)
    public var bankName: String                 // "State Bank of India"
    public var provisional: Bool                // shape confirmed, quirks unverified
    /// role → case-insensitive regex matched against a header cell.
    /// Roles: date, narration, reference, debit, credit, amount, drcr, balance.
    public var headerPatterns: [String: String]
    /// Whole-row regexes (cells joined with "|") dropped from the body.
    public var furniturePatterns: [String]
    /// Tried in order; the first that parses every body date wins.
    public var dateFormats: [String]
    /// "debitCredit" | "signedAmount" | "amountDRCR" | "unsignedChain"
    public var signConvention: String
    /// Optional regexes (one capture group) extracting a rail reference from
    /// the narration when no reference cell is present. See ColumnMapping.
    public var referencePatterns: [String]?

    public init(id: String, sourceID: String, bankName: String, provisional: Bool = false,
                headerPatterns: [String: String], furniturePatterns: [String] = [],
                dateFormats: [String], signConvention: String,
                referencePatterns: [String]? = nil) {
        self.id = id
        self.sourceID = sourceID
        self.bankName = bankName
        self.provisional = provisional
        self.headerPatterns = headerPatterns
        self.furniturePatterns = furniturePatterns
        self.dateFormats = dateFormats
        self.signConvention = signConvention
        self.referencePatterns = referencePatterns
    }
}

public enum SpecStore {
    /// Decodes every Resources/formats/*.json in the package bundle.
    public static func bundled() -> [FormatSpec] {
        guard let urls = Bundle.module.urls(forResourcesWithExtension: "json",
                                            subdirectory: "Resources/formats") else { return [] }
        let decoder = JSONDecoder()
        var specs: [FormatSpec] = []
        for url in urls.sorted(by: { $0.lastPathComponent < $1.lastPathComponent }) {
            if let data = try? Data(contentsOf: url),
               let spec = try? decoder.decode(FormatSpec.self, from: data) {
                specs.append(spec)
            }
        }
        return specs
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/generic_bank/format_spec.dart
/// Declarative bank-statement layout. Port of FormatSpec.swift. Hard rule:
/// this format never grows an expression language or any eval-like field.
library;

import 'dart:convert';

class FormatSpec {
  final String id;
  final String sourceID;
  final String bankName;
  final bool provisional;
  /// role → case-insensitive regex matched against a header cell.
  /// Roles: date, narration, reference, debit, credit, amount, drcr, balance.
  final Map<String, String> headerPatterns;
  /// Whole-row regexes (cells joined with "|") dropped from the body.
  final List<String> furniturePatterns;
  /// Tried in order; the first that parses every body date wins.
  final List<String> dateFormats;
  /// "debitCredit" | "signedAmount" | "amountDRCR" | "unsignedChain"
  final String signConvention;
  final List<String>? referencePatterns;
  /// Case-insensitive regexes that must each match somewhere in the document
  /// text for the spec to apply (disambiguates identical table shapes).
  final List<String>? detectPatterns;

  const FormatSpec({
    required this.id,
    required this.sourceID,
    required this.bankName,
    this.provisional = false,
    required this.headerPatterns,
    this.furniturePatterns = const [],
    required this.dateFormats,
    required this.signConvention,
    this.referencePatterns,
    this.detectPatterns,
  });

  factory FormatSpec.fromJsonString(String json) {
    final map = jsonDecode(json) as Map<String, dynamic>;
    return FormatSpec(
      id: map['id'] as String,
      sourceID: map['sourceID'] as String,
      bankName: map['bankName'] as String,
      provisional: map['provisional'] as bool? ?? false,
      headerPatterns: (map['headerPatterns'] as Map<String, dynamic>)
          .map((k, v) => MapEntry(k, v as String)),
      furniturePatterns: (map['furniturePatterns'] as List? ?? [])
          .map((e) => e as String)
          .toList(),
      dateFormats:
          (map['dateFormats'] as List).map((e) => e as String).toList(),
      signConvention: map['signConvention'] as String,
      referencePatterns: (map['referencePatterns'] as List?)
          ?.map((e) => e as String)
          .toList(),
      detectPatterns: (map['detectPatterns'] as List?)
          ?.map((e) => e as String)
          .toList(),
    );
  }
}

class SpecStore {
  /// Parses a set of spec JSON strings (the app supplies asset contents; the
  /// package stays IO-free). Silently drops undecodable entries — CI's
  /// bundled-spec tests catch those upstream.
  static List<FormatSpec> parseAll(Iterable<String> jsonStrings) {
    final specs = <FormatSpec>[];
    for (final json in jsonStrings) {
      try {
        specs.add(FormatSpec.fromJsonString(json));
      } catch (_) {
        continue;
      }
    }
    specs.sort((a, b) => a.id.compareTo(b.id));
    return specs;
  }
}
